File: SQL Server 2012/Exercises/Chapter 22/Visual Basic/ApClrObjectsEx/GetLineItemCountDateRange.vb
Imports System
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.SqlTypes
Imports Microsoft.SqlServer.Server


Partial Public Class StoredProcedures
    <Microsoft.SqlServer.Server.SqlProcedure()> _
    Public Shared Sub GetLineItemCountDateRange(
            ByRef lineItemCount As SqlInt32,
            ByVal startDate As SqlDateTime,
            ByVal endDate As SqlDateTime)
        If startDate.IsNull Then
            Throw New ApplicationException("Invalid start date. " &
                "You must specify a start date.")
        End If
        If startDate > DateTime.Today Then
            Throw New ApplicationException("Invalid start date. " &
                "The start date can't be in the future.")
        End If
        If endDate.IsNull Then
            endDate = DateTime.Today
        End If
        If endDate < startDate Then
            Throw New ApplicationException("Invalid end date. " &
                "The end date can't be before the start date.")
        End If
        Dim connection As New SqlConnection("Context connection=true")
        connection.Open()

        Dim sql As String =
            "SELECT COUNT(*) " &
            "FROM InvoiceLineItems JOIN Invoices " &
            "ON InvoiceLineItems.InvoiceID = Invoices.InvoiceID " &
            "WHERE ((InvoiceDate >= @StartDate) AND " &
            "       (InvoiceDate <= @EndDate))"

        Dim command As New SqlCommand(sql, connection)

        Dim param As New SqlParameter("@StartDate", SqlDbType.DateTime)
        param.Value = startDate
        command.Parameters.Add(param)

        param = New SqlParameter("@EndDate", SqlDbType.DateTime)
        param.Value = endDate
        command.Parameters.Add(param)

        Try
            Dim scalarValue As Integer = CInt(command.ExecuteScalar)
            lineItemCount = CType(scalarValue, SqlInt32)
        Catch ex As Exception
            lineItemCount = 0
        End Try

        connection.Close()
    End Sub
End Class

File: SQL Server 2012/Exercises/Chapter 22/C#/ApClrObjectsEx/ApClrObjectsEx/GetLineItemCountDateRange.cs
//------------------------------------------------------------------------------
// <copyright file="CSSqlStoredProcedure.cs" company="Microsoft">
//     Copyright (c) Microsoft Corporation.  All rights reserved.
// </copyright>
//------------------------------------------------------------------------------
using System;
using System.Data;
using System.Data.SqlClient;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;

public partial class StoredProcedures
{
    [Microsoft.SqlServer.Server.SqlProcedure]
    public static void GetLineItemCountDateRange (
        out SqlInt32 lineItemCount,
        SqlDateTime startDate,
        SqlDateTime endDate)
    {
        if (startDate.IsNull)
            throw new ApplicationException("Invalid start date. " +
                "You must specify a start date.");
        if (startDate > DateTime.Today)
            throw new ApplicationException("Invalid start date. " +
                "The start date can't be in the future.");
        if (endDate.IsNull)
            endDate = DateTime.Today;
        if (endDate < startDate)
            throw new ApplicationException("Invalid end date. " +
                "The end date can't be before the start date.");
        SqlConnection connection = new SqlConnection("Context connection=true");
        connection.Open();

        String sql =
            "SELECT COUNT(*) " +
            "FROM InvoiceLineItems JOIN Invoices " +
            "ON InvoiceLineItems.InvoiceID = Invoices.InvoiceID " +
            "WHERE ((InvoiceDate >= @StartDate) AND " +
            "       (InvoiceDate <= @EndDate))";

        SqlCommand command = new SqlCommand(sql, connection);

        SqlParameter param = new SqlParameter("@StartDate", SqlDbType.DateTime);
        param.Value = startDate;
        command.Parameters.Add(param);

        param = new SqlParameter("@EndDate", SqlDbType.DateTime);
        param.Value = endDate;
        command.Parameters.Add(param);

        try
        {
            int scalarValue = (int)command.ExecuteScalar();
            lineItemCount = (SqlInt32)scalarValue;
        }
        catch
        {
            lineItemCount = 0;
        }

        connection.Close();
    }
}
